affinity/cpp-23/dXXXXr0.md

```diff
@@ -0,0 +1,295 @@
+accesses to remote locations incur higher latency and/or lower bandwidth,
+control of thread placement to enforce affinity within parallel applications is
+crucial to fuel all the cores and to exploit the full performance of the memory
+subsystem on NUMA computers. 
+
+Operating systems (OSes) traditionally take responsibility for assigning threads
+or processes to run on processing units. However, OSes may use high-level
+policies for this assignment that do not necessarily match the optimal usage
+pattern for a given application. Application developers must leverage the
+placement of memory and *placement of threads* for best performance on current
+and future architectures. For C\+\+ developers to achieve this, native support
+for *placement of threads and memory* is critical for application portability.
+We will refer to this as the *affinity problem*. 
+
+The affinity problem is especially challenging for applications whose behavior
+changes over time or is hard to predict, or when different applications
+interfere with each other's performance. Today, most OSes already can group
+processing units according to their locality and distribute processes, while
+keeping threads close to the initial thread, or even avoid migrating threads and
+maintain first touch policy. Nevertheless, most programs can change their work
+distribution, especially in the presence of nested parallelism.
+
+Frequently, data are initialized at the beginning of the program by the initial
+thread and are used by multiple threads. While some OSes automatically migrate
+threads or data for better affinity, migration may have high overhead. In an
+optimal case, the OS may automatically detect which thread access which data
+most frequently, or it may replicate data which are read by multiple threads, or
+migrate data which are modified and used by threads residing on remote locality
+groups. However, the OS often does a reasonable job, if the machine is not
+overloaded, if the application carefully uses first-touch allocation, and if the
+program does not change its behavior with respect to locality.
+
+The affinity interface we propose should help computers achieve a much higher
+fraction of peak memory bandwidth when using parallel algorithms. In the future,
+we plan to extend this to heterogeneous and distributed computing. This follows
+the lead of OpenMP [[3]][design-of-openmp].
+
+# Proposal
+
+This paper proposes a new group of properties (as described in P1393r0
+[[4]][p1393r0] that allow users to specify how a callable in a bulk algorithm
+such as the `bulk_execute` algorithm in [[2]][p0443r12] will perform in light of
+the placement of the work-items it invokes. This information can then be used by
+an an executor implementation which is aware of the property to efficiently map
+the work-items it invokes to the underlying resources which it represents.
+
+## Shift to descriptive properties
+
+The previous revision of this proposal; P1436r3 [[1]][p1436r3] took a
```

Comment:
"The previous revision of this proposal, P1463r3 [[1]][p1436r3], took a ...."

Reply:
Done.